ops/model.py

```diff
@@ -600,6 +602,48 @@ def add_secret(self, content: Dict[str, str], *,
             owner='unit')
         return Secret(self._backend, id=id, label=label, content=content)
 
+    def open_port(self, protocol: typing.Literal['tcp', 'udp', 'icmp'],
+                  port: Optional[int] = None):
+        """Open a port with the given protocol for this unit.
+
+        Calling this registers intent with Juju that the application should be
+        accessed on the given port, but the port isn't actually opened
+        externally until the operator runs "juju expose".
+
+        On Kubernetes sidecar charms, the ports opened are not strictly
+        per-unit: Juju will open the union of ports from all units.
+        However, normally charms should make the same open_port() call from
+        every unit.
+
+        Args:
+            protocol: String representing the protocol; must be one of
+                'tcp', 'udp', or 'icmp' (lowercase is recommended, but
+                uppercase is also supported).
+            port: The port to open. Required for TCP and UDP; not allowed
+                for ICMP.
+        """
```

Comment:
Do we document in the docstring here what the backend may raise?

Reply:
We don't usually for all of these when they just raise `ModelError` (what all hook tool errors end up as). If they raise more specific exceptions we document that.

I'd be open to reconsider that and also document the `Raises: ModelError ...`, but it would be on pretty much every function.

Reply:
Nice, I didn't realize all hook tools only raise ModelError.